tests/gradients/core/test_pulse_generator_gradient.py

```diff
@@ -693,3 +703,621 @@ def _int_sin(p, t):
+        _match = "gradient of variances with the pulse generator parameter-shift gradient"
+        with pytest.raises(ValueError, match=_match):
+            pulse_generator(tape)
+
+    def test_raises_with_invalid_op(self):
+        """Test that an error is raised when calling ``pulse_generator`` on a non-pulse op."""
+        tape = qml.tape.QuantumScript([qml.RX(0.4, 0)], [qml.expval(Z(0))])
+        _match = "pulse_generator does not support differentiating parameters of other"
+        with pytest.raises(ValueError, match=_match):
+            pulse_generator(tape)
+
+    def test_no_trainable_params_tape(self):
+        """Test that the correct ouput and warning is generated in the absence of any trainable
+        parameters"""
+        dev = qml.device("default.qubit", wires=2)
+
+        weights = [0.1, 0.2]
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=0)
+            qml.expval(Z(0) @ Z(1))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        # TODO: remove once #2155 is resolved
+        tape.trainable_params = []
+
+        with pytest.warns(UserWarning, match="gradient of a tape with no trainable parameters"):
+            g_tapes, post_processing = pulse_generator(tape)
+        res = post_processing(qml.execute(g_tapes, dev, None))
+
+        assert g_tapes == []
+        assert isinstance(res, np.ndarray)
+        assert res.shape == (0,)
+
+    def test_no_trainable_params_multiple_return_tape(self):
+        """Test that the correct ouput and warning is generated in the absence of any trainable
+        parameters with multiple returns."""
+        dev = qml.device("default.qubit", wires=2)
+
+        weights = [0.1, 0.2]
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(weights[0], wires=0)
+            qml.RY(weights[1], wires=0)
+            qml.expval(Z(0) @ Z(1))
+            qml.probs(wires=[0, 1])
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        tape.trainable_params = []
+
+        with pytest.warns(UserWarning, match="gradient of a tape with no trainable parameters"):
+            _tapes, fn = pulse_generator(tape)
+        res = fn(dev.batch_execute(_tapes))
+
+        assert _tapes == []
+        assert isinstance(res, tuple)
+        for r in res:
+            assert isinstance(r, np.ndarray)
+            assert r.shape == (0,)
+
+    def test_all_zero_diff_methods_tape(self):
+        """Test that the transform works correctly when the diff method for every parameter is
+        identified to be 0, and that no tapes were generated."""
+        dev = qml.device("default.qubit", wires=3)
+
+        params = pnp.array([0.5, 0.5, 0.5], requires_grad=True)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.Rot(*params, wires=0)
+            qml.probs([2, 3])
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+
+        res_pulse_gen, _tapes = grad_fn(tape, dev)
```

Comment:
- The operation will have a zero gradient, as determined by `gradient_analysis_and_validation`, because it acts on a wire that has no impact on the measurement at all.
- `_assert_has_jax` only checks whether `JAX` is installed.
- The check for operations being a `ParametrizedEvolution` is made further down the line, in `_generate_tapes_and_coeffs`. This makes sense because a non-pules operation in a tape is fine as long as we don't want to tackle it with the `pulse_generator` method, or it does not affect the measurements as in the test example :)

Reply:
Oh, that all makes sense. Thanks for clarifying!